tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/localedistance/LocaleDistanceMapper.java

```diff
@@ -0,0 +1,465 @@
+// © 2017 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html#License
+package org.unicode.icu.tool.cldrtoicu.localedistance;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Arrays.asList;
+import static org.unicode.cldr.api.CldrData.PathOrder.DTD;
+import static org.unicode.cldr.api.CldrDataType.SUPPLEMENTAL;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.unicode.cldr.api.AttributeKey;
+import org.unicode.cldr.api.CldrData;
+import org.unicode.cldr.api.CldrData.ValueVisitor;
+import org.unicode.cldr.api.CldrDataSupplier;
+import org.unicode.cldr.api.CldrPath;
+import org.unicode.cldr.api.CldrValue;
+import org.unicode.cldr.api.PathMatcher;
+import org.unicode.icu.tool.cldrtoicu.IcuData;
+import org.unicode.icu.tool.cldrtoicu.RbPath;
+import org.unicode.icu.tool.cldrtoicu.RbValue;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.primitives.Bytes;
+import com.ibm.icu.impl.locale.LSR;
+import com.ibm.icu.impl.locale.LocaleDistance;
+import com.ibm.icu.impl.locale.XLikelySubtags;
+import com.ibm.icu.util.ULocale;
+
+/**
+ * Mapper for generating locale distance tables from CLDR language data.
+ *
+ * <p>Note that this is an atypical mapper which does a lot more processing than other
+ * ICU mapper classes and relies on several auxilliary classes (which is why it's in a
+ * different package). Conceptually it's still a "mapper" though, just not a simple one.
+ *
+ * <p>This mapper was converted from the LocaleDistanceBuilder code in the ICU4J project,
+ * but not all aspects of it have been understood completely, so documentation may be a
+ * bit lacking.
+ */
+public final class LocaleDistanceMapper {
```

Comment:
The main top level mapper. The class is now simply structured and contains almost no complexity (~4 'if' statements in the entire class). It drives the conversion by deferring to a series of well defined (and mostly separable) steps, each generating a well defined output.